exp/internal/controllers/machinepool_controller_test.go

```diff
@@ -336,82 +342,141 @@ func TestReconcileMachinePoolRequest(t *testing.T) {
-				err:    false,
+				mpExist: true,
+				result:  reconcile.Result{},
+				err:     false,
 			},
+			withTracker: true,
 		},
 		{
+			name: "Delete machinePool with nodeDeleteTimeout passed",
 			machinePool: expv1.MachinePool{
 				ObjectMeta: metav1.ObjectMeta{
-					Name:      "deleted",
+					Name:      "deleted-timeoutNode",
 					Namespace: metav1.NamespaceDefault,
 					Labels: map[string]string{
 						clusterv1.MachineControlPlaneLabel: "",
 					},
 					Finalizers:        []string{expv1.MachinePoolFinalizer},
-					DeletionTimestamp: &time,
+					CreationTimestamp: metav1.Time{Time: timeNow.Add(time.Minute * -2)},
+					DeletionTimestamp: &metav1.Time{Time: timeNow.Add(time.Minute * -1)},
 				},
 				Spec: expv1.MachinePoolSpec{
 					ClusterName: "test-cluster",
 					Replicas:    ptr.To[int32](1),
 					Template: clusterv1.MachineTemplateSpec{
 						Spec: clusterv1.MachineSpec{
-							InfrastructureRef: corev1.ObjectReference{
-								APIVersion: builder.InfrastructureGroupVersion.String(),
-								Kind:       builder.TestInfrastructureMachineTemplateKind,
-								Name:       "infra-config1",
-							},
-							Bootstrap: clusterv1.Bootstrap{DataSecretName: ptr.To("data")},
+							NodeDeletionTimeout: &metav1.Duration{Duration: 10 * time.Second},
+						},
+					},
+					ProviderIDList: []string{"aws:///us-test-2a/i-013ab00756982217f"},
+				},
+				Status: expv1.MachinePoolStatus{
+					NodeRefs: []corev1.ObjectReference{
+						{
+							APIVersion: "v1",
+							Kind:       "Node",
+							Name:       "test-node",
 						},
 					},
 				},
 			},
 			expected: expected{
-				result: reconcile.Result{},
-				err:    false,
+				mpExist: false,
+				result:  reconcile.Result{},
+				err:     true,
+			},
+		},
+		{
+			name: "Try delete machinePool with nodeDeleteTimeout set to zero",
+			machinePool: expv1.MachinePool{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "try-delete",
+					Namespace: metav1.NamespaceDefault,
+					Labels: map[string]string{
+						clusterv1.MachineControlPlaneLabel: "",
+					},
+					Finalizers:        []string{expv1.MachinePoolFinalizer},
+					CreationTimestamp: metav1.Time{Time: timeNow.Add(time.Minute * -2)},
+					DeletionTimestamp: &metav1.Time{Time: timeNow.Add(time.Minute * -1)},
+				},
+				Spec: expv1.MachinePoolSpec{
+					ClusterName: "test-cluster",
+					Replicas:    ptr.To[int32](1),
+					Template: clusterv1.MachineTemplateSpec{
+						Spec: clusterv1.MachineSpec{
+							NodeDeletionTimeout: &metav1.Duration{Duration: 0 * time.Second},
+						},
+					},
+					ProviderIDList: []string{"aws:///us-test-2a/i-013ab00756982217f"},
+				},
+				Status: expv1.MachinePoolStatus{
+					NodeRefs: []corev1.ObjectReference{
+						{
+							APIVersion: "v1",
+							Kind:       "Node",
+							Name:       "test-node",
+						},
+					},
+				},
+			},
+			expected: expected{
+				mpExist: true,
+				result:  reconcile.Result{},
+				err:     true,
 			},
```

Comment:
```suggestion
			},
			withTracker: true,
```
Otherwise we are always hitting the code path that the tracker is nil. I think that his not what we want to test here

Reply:
Actually let's please drop the withTracker field and let's just always set the tracker. I don't see a reason why we should test the case where Tracker is not set, as this never happens at runtime

Reply:
We set here the tracker to nil cause we don't want execute [delete_nodes.](https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/internal/controllers/machinepool_controller.go#L283) Otherwise the node is not retrievable and it will be ignored [ignore_nodes](https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/internal/controllers/machinepool_controller_noderef.go#L136). We want to be sure when nodeTimeout set to zero the machinePool will continues try to delete

Reply:
So this test tests that we are getting errors when the Tracker is not set? Which is something that never happens in reality. I would be much more comfortable with our test coverage if we would test real cases instead of something like this

Reply:
It should be possible to create a few Nodes with some finalizers that block deletion in the fake cluster cache tracker